File: daemon/lib/api/storage_api.dart
import 'dart:convert';
import 'package:shelf/shelf.dart' as shelf;
import 'package:shelf_router/shelf_router.dart';
import 'package:opencli_daemon/database/app_database.dart';

/// REST API endpoints for client-side storage (web UI, Flutter app).
///
/// Provides CRUD for generation history, assets, status events, and chat messages.
/// All data is persisted in the centralized SQLite database.
class StorageApi {
  final AppDatabase _db;

  StorageApi({AppDatabase? db}) : _db = db ?? AppDatabase.instance;

  /// Register all storage routes on the given router.
  void registerRoutes(Router router) {
    // Generation History
    router.get('/api/v1/history', _listHistory);
    router.post('/api/v1/history', _createHistory);
    router.delete('/api/v1/history/<id>', _deleteHistory);
    router.delete('/api/v1/history', _clearHistory);

    // Assets
    router.get('/api/v1/assets', _listAssets);
    router.post('/api/v1/assets', _createAsset);
    router.delete('/api/v1/assets/<id>', _deleteAsset);

    // Status Events
    router.get('/api/v1/events', _listEvents);
    router.post('/api/v1/events', _createEvent);
    router.get('/api/v1/events/stats', _getEventStats);

    // Chat Messages
    router.get('/api/v1/chat/messages', _listChatMessages);
    router.post('/api/v1/chat/messages', _createChatMessage);
    router.delete('/api/v1/chat/messages', _clearChatMessages);
  }

  // ── Generation History ──

  Future<shelf.Response> _listHistory(shelf.Request request) async {
    try {
      final limit =
          int.tryParse(request.url.queryParameters['limit'] ?? '') ?? 50;
      final rows = await _db.listHistory(limit: limit);
      return _json({'history': rows});
    } catch (e) {
      return _error('Failed to list history: $e');
    }
  }

  Future<shelf.Response> _createHistory(shelf.Request request) async {
    try {
      final body = jsonDecode(await request.readAsString());
      final now = DateTime.now().millisecondsSinceEpoch;
      await _db.insertHistory({
        'id': body['id'] ?? 'h_$now',
        'mode': body['mode'] ?? '',
        'prompt': body['prompt'] ?? '',
        'provider': body['provider'] ?? '',
        'style': body['style'] ?? '',
        'result_type': body['result_type'] ?? body['resultType'] ?? '',
        'thumbnail': body['thumbnail'],
        'created_at': body['created_at'] ?? body['timestamp'] ?? now,
      });
      return _json({'success': true});
    } catch (e) {
      return _error('Failed to create history: $e');
    }
  }

  Future<shelf.Response> _deleteHistory(
      shelf.Request request, String id) async {
    try {
      await _db.deleteHistory(id);
      return _json({'success': true});
    } catch (e) {
      return _error('Failed to delete history: $e');
    }
  }

  Future<shelf.Response> _clearHistory(shelf.Request request) async {
    try {
      await _db.clearHistory();
      return _json({'success': true});
    } catch (e) {
      return _error('Failed to clear history: $e');
    }
  }

  // ── Assets ──

  Future<shelf.Response> _listAssets(shelf.Request request) async {
    try {
      final limit =
          int.tryParse(request.url.queryParameters['limit'] ?? '') ?? 100;
      final rows = await _db.listAssets(limit: limit);
      return _json({'assets': rows});
    } catch (e) {
      return _error('Failed to list assets: $e');
    }
  }

  Future<shelf.Response> _createAsset(shelf.Request request) async {
    try {
      final body = jsonDecode(await request.readAsString());
      final now = DateTime.now().millisecondsSinceEpoch;
      await _db.insertAsset({
        'id': body['id'] ?? 'asset_$now',
        'type': body['type'] ?? 'image',
        'title': body['title'] ?? '',
        'url': body['url'] ?? '',
        'thumbnail': body['thumbnail'],
        'provider': body['provider'],
        'style': body['style'],
        'created_at': body['created_at'] ?? body['createdAt'] ?? now,
      });
      return _json({'success': true});
    } catch (e) {
      return _error('Failed to create asset: $e');
    }
  }

  Future<shelf.Response> _deleteAsset(
      shelf.Request request, String id) async {
    try {
      await _db.deleteAsset(id);
      return _json({'success': true});
    } catch (e) {
      return _error('Failed to delete asset: $e');
    }
  }

  // ── Status Events ──

  Future<shelf.Response> _listEvents(shelf.Request request) async {
    try {
      final limit =
          int.tryParse(request.url.queryParameters['limit'] ?? '') ?? 100;
      final rows = await _db.listEvents(limit: limit);
      return _json({'events': rows});
    } catch (e) {
      return _error('Failed to list events: $e');
    }
  }

  Future<shelf.Response> _createEvent(shelf.Request request) async {
    try {
      final body = jsonDecode(await request.readAsString());
      final now = DateTime.now().millisecondsSinceEpoch;
      await _db.insertEvent({
        'id': body['id'] ?? 'evt_$now',
        'type': body['type'] ?? 'system',
        'source': body['source'] ?? '',
        'content': body['content'] ?? '',
        'task_type': body['task_type'] ?? body['taskType'],
        'status': body['status'],
        'result': body['result'] is String
            ? body['result']
            : (body['result'] != null ? jsonEncode(body['result']) : null),
        'created_at': body['created_at'] ?? body['timestamp'] ?? now,
      });
      return _json({'success': true});
    } catch (e) {
      return _error('Failed to create event: $e');
    }
  }

  Future<shelf.Response> _getEventStats(shelf.Request request) async {
    try {
      final stats = await _db.getEventStats();
      return _json(stats);
    } catch (e) {
      return _error('Failed to get event stats: $e');
    }
  }

  // ── Chat Messages ──

  Future<shelf.Response> _listChatMessages(shelf.Request request) async {
    try {
      final limit =
          int.tryParse(request.url.queryParameters['limit'] ?? '') ?? 100;
      final rows = await _db.listChatMessages(limit: limit);
      return _json({'messages': rows});
    } catch (e) {
      return _error('Failed to list chat messages: $e');
    }
  }

  Future<shelf.Response> _createChatMessage(shelf.Request request) async {
    try {
      final body = jsonDecode(await request.readAsString());
      final now = DateTime.now().millisecondsSinceEpoch;
      await _db.insertChatMessage({
        'id': body['id'] ?? 'msg_$now',
        'content': body['content'] ?? '',
        'is_user': (body['is_user'] ?? body['isUser'] ?? false) ? 1 : 0,
        'timestamp': body['timestamp'] ?? now,
        'status': body['status'] ?? 'completed',
        'task_type': body['task_type'] ?? body['taskType'],
        'result': body['result'] is String
            ? body['result']
            : (body['result'] != null ? jsonEncode(body['result']) : null),
      });
      return _json({'success': true});
    } catch (e) {
      return _error('Failed to create chat message: $e');
    }
  }

  Future<shelf.Response> _clearChatMessages(shelf.Request request) async {
    try {
      await _db.clearChatMessages();
      return _json({'success': true});
    } catch (e) {
      return _error('Failed to clear chat messages: $e');
    }
  }

  // ── Helpers ──

  shelf.Response _json(Map<String, dynamic> data) {
    return shelf.Response.ok(
      jsonEncode(data),
      headers: {'Content-Type': 'application/json'},
    );
  }

  shelf.Response _error(String message) {
    return shelf.Response.internalServerError(
      body: jsonEncode({'error': message}),
      headers: {'Content-Type': 'application/json'},
    );
  }
}

File: daemon-py/opencli_daemon/api/storage_api.py
"""Storage REST API — history, assets, events, chat messages.

Ported from daemon/lib/api/storage_api.dart.
"""

import json
import time
from typing import Any

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from opencli_daemon.database import connection as db

router = APIRouter(prefix="/api/v1", tags=["storage"])


def _now() -> int:
    return int(time.time() * 1000)


# ── Generation History ───────────────────────────────────────────────────────


@router.get("/history")
async def list_history(limit: int = 50) -> dict:
    rows = await db.list_rows("generation_history", order_by="created_at DESC", limit=limit)
    return {"history": rows}


@router.post("/history")
async def create_history(request: Request) -> dict:
    body = await request.json()
    now = _now()
    await db.insert_capped("generation_history", {
        "id": body.get("id", f"h_{now}"),
        "mode": body.get("mode", ""),
        "prompt": body.get("prompt", ""),
        "provider": body.get("provider", ""),
        "style": body.get("style", ""),
        "result_type": body.get("result_type") or body.get("resultType", ""),
        "thumbnail": body.get("thumbnail"),
        "created_at": body.get("created_at") or body.get("timestamp", now),
    }, max_rows=50)
    return {"success": True}


@router.delete("/history/{item_id}")
async def delete_history(item_id: str) -> dict:
    await db.delete_row("generation_history", "id", item_id)
    return {"success": True}


@router.delete("/history")
async def clear_history() -> dict:
    await db.delete_all("generation_history")
    return {"success": True}


# ── Assets ───────────────────────────────────────────────────────────────────


@router.get("/assets")
async def list_assets(limit: int = 100) -> dict:
    rows = await db.list_rows("assets", order_by="created_at DESC", limit=limit)
    return {"assets": rows}


@router.post("/assets")
async def create_asset(request: Request) -> dict:
    body = await request.json()
    now = _now()
    await db.insert_capped("assets", {
        "id": body.get("id", f"asset_{now}"),
        "type": body.get("type", "image"),
        "title": body.get("title", ""),
        "url": body.get("url", ""),
        "thumbnail": body.get("thumbnail"),
        "provider": body.get("provider"),
        "style": body.get("style"),
        "created_at": body.get("created_at") or body.get("createdAt", now),
    }, max_rows=100)
    return {"success": True}


@router.delete("/assets/{item_id}")
async def delete_asset(item_id: str) -> dict:
    await db.delete_row("assets", "id", item_id)
    return {"success": True}


# ── Status Events ────────────────────────────────────────────────────────────


@router.get("/events")
async def list_events(limit: int = 100) -> dict:
    rows = await db.list_rows("status_events", order_by="created_at DESC", limit=limit)
    return {"events": rows}


@router.post("/events")
async def create_event(request: Request) -> dict:
    body = await request.json()
    now = _now()
    result = body.get("result")
    if result is not None and not isinstance(result, str):
        result = json.dumps(result)
    await db.insert_capped("status_events", {
        "id": body.get("id", f"evt_{now}"),
        "type": body.get("type", "system"),
        "source": body.get("source", ""),
        "content": body.get("content", ""),
        "task_type": body.get("task_type") or body.get("taskType"),
        "status": body.get("status"),
        "result": result,
        "created_at": body.get("created_at") or body.get("timestamp", now),
    }, max_rows=500)
    return {"success": True}


@router.get("/events/stats")
async def get_event_stats() -> dict:
    total = await db.count_rows("status_events")
    completed = await db.count_rows("status_events", "status = ?", ("completed",))
    failed = await db.count_rows("status_events", "status = ?", ("failed",))
    one_min_ago = _now() - 60_000
    recent = await db.count_rows("status_events", "created_at > ?", (one_min_ago,))
    denom = completed + failed
    success_rate = completed / denom if denom > 0 else 1.0
    return {
        "total": total,
        "completed": completed,
        "failed": failed,
        "success_rate": success_rate,
        "tasks_per_min": recent,
    }


# ── Chat Messages ────────────────────────────────────────────────────────────


@router.get("/chat-messages")
async def list_chat_messages_compat(limit: int = 100) -> dict:
    """Compat alias for /chat/messages."""
    return await list_chat_messages(limit)


@router.get("/chat/messages")
async def list_chat_messages(limit: int = 100) -> dict:
    rows = await db.list_rows("chat_messages", order_by="timestamp DESC", limit=limit)
    return {"messages": rows}


@router.post("/chat/messages")
async def create_chat_message(request: Request) -> dict:
    body = await request.json()
    now = _now()
    result = body.get("result")
    if result is not None and not isinstance(result, str):
        result = json.dumps(result)
    is_user = body.get("is_user") or body.get("isUser", False)
    await db.insert_capped("chat_messages", {
        "id": body.get("id", f"msg_{now}"),
        "content": body.get("content", ""),
        "is_user": 1 if is_user else 0,
        "timestamp": body.get("timestamp", now),
        "status": body.get("status", "completed"),
        "task_type": body.get("task_type") or body.get("taskType"),
        "result": result,
    }, max_rows=100, order_col="timestamp")
    return {"success": True}


@router.delete("/chat/messages")
async def clear_chat_messages() -> dict:
    await db.delete_all("chat_messages")
    return {"success": True}
